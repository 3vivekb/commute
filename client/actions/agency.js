import fetchAction from '@conveyal/woonerf/fetch'
import {push} from 'react-router-redux'
import {createAction} from 'redux-actions'
import uuid from 'uuid'

const addLocally = createAction('add agency')
export const createAgency = (newAgency) => {
  newAgency.id = uuid.v4()
  newAgency.organizations = []
  return [
    addLocally(newAgency),
    push(`/agency/${newAgency.id}`)
  ] // TODO save to server
}

const deleteLocally = createAction('delete agency')
/* const deleteOnServer = (id) =>
  serverAction({
    url: `/api/agency/${id}`,
    options: {
      method: 'delete'
    }
  }) */ // TODO delete on server
export const deleteAgency = (id) => [deleteLocally(id), push('/')]

export const loadAgencies = () => fetchAction({
  url: '/api/agency',
  next: (res, err) => {
    if (!err) {
      return setAllLocally(res)
    }
  }
})

export const updateAgency = (agency) => [
  setLocally(agency),
  push(`/agency/${agency.id}`)
]
/* const updateOnServer = (agency) =>
  serverAction({
    url: `/api/agency/${agency.id}`,
    options: {
      body: JSON.stringify(agency),
      method: 'put'
    }
  }) */  // TODO: update on server

export const setLocally = createAction('set agency')
export const setAllLocally = createAction('set agencies')
