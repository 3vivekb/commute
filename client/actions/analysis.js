import moment from 'moment'
import {push} from 'react-router-redux'
import {createAction} from 'redux-actions'
import uuid from 'uuid'

const addAnalysis = createAction('add analysis')
export const createAnalysis = (newAnalysis, organizationId) => {
  newAnalysis.id = uuid.v4()
  newAnalysis.lastRunDateTime = moment().unix()
  newAnalysis.trips = []
  return [
    addAnalysis({ analysis: newAnalysis, organizationId }),
    push(`/organizations/${organizationId}/analysis/${newAnalysis.id}`)
  ] // TODO save to server
}

const deleteLocally = createAction('delete analysis')
/* const deleteOnServer = (id) =>
  serverAction({
    url: `/api/analysis/${id}`,
    options: {
      method: 'delete'
    }
  }) */ // TODO delete on server
export const deleteAnalysis = (id, organzationId) => [
  deleteLocally({ analysisId: id, organzationId }),
  push(`/organizations/${organzationId}`)
]
