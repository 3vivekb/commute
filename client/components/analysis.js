import React, {Component} from 'react'

export default class Analysis extends Component {
  static propTypes = {}

  render () {
    return <h1>Analysis</h1>
  }
}
