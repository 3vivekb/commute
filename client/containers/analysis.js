import {connect} from 'react-redux'

import analysisActions from '../actions/analysis'
import _histogram from '../components/analysis/histogram'
import _individuals from '../components/analysis/individuals'
import _possibilities from '../components/analysis/possibilities'
import _summary from '../components/analysis/summary'

function mapStateToProps (state, props) {
  const {analysis, commuter, group, site} = state
  const currentAnalysis = analysis[props.params.analysisId]
  return {
    analysis: currentAnalysis,
    commuterStore: commuter,
    groupName: group[currentAnalysis.groupId].name,
    siteName: site[currentAnalysis.siteId].name
  }
}

function mapDispatchToProps (dispatch) {
  return {
    deleteAnalysis: (opts) => dispatch(analysisActions.delete(opts)),
    loadAnalysis: (opts) => dispatch(analysisActions.loadOne(opts))
  }
}

export const Histogram = connect(mapStateToProps)(_histogram)
export const Individuals = connect(mapStateToProps)(_individuals)
export const Possibilities = connect(mapStateToProps)(_possibilities)
export const Summary = connect(mapStateToProps, mapDispatchToProps)(_summary)
