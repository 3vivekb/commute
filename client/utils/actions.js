import fetchAction from '@conveyal/woonerf/fetch'
import qs from 'qs'
import {push} from 'react-router-redux'
import {createAction} from 'redux-actions'

/**
 * Make generic model actions that communicate with a rest server
 *
 * @param  {Object} cfg An object with the following parameters:
 * - commands: An object with the keys being commands and the corresponding cfg of each command
 * -- redirectionStrategy: The redirection strategy to use upon success.
 *  Can be one of following: 'toEntity', 'toParent', 'toHome', undefined (no redirection)
 *  Defaults are as follows:
 *    Collection GET, GET: null
 *    Collection POST, PUT: toEntity
 *    DELETE: toParent
 * - parentKey: Key in model of parent's id
 *   (must be provided if using toParent redirection strategy)
 * - parentName: Name of app route to parent view
 *   (must be provided if using toParent redirection strategy)
 * - pluralName: Plural name of the model
 * - singularName: Singular name of the model
 * @return {Object}     The actions
 */
const makeGenericModelActions = (cfg) => {
  const {commands, parentKey, parentName, pluralName, singularName} = cfg
  const actions = {}

  // make local set actions
  const addLocally = createAction(`add ${singularName}`)
  const deleteLocally = createAction(`delete ${singularName}`)
  const setLocally = createAction(`set ${singularName}`)
  const setAllLocally = createAction(`set ${pluralName}`)

  const baseEndpoint = `/api/${singularName}`

  const redirectionStrategies = {
    'toEntity': (entity) => {
      return push(`/${singularName}/${entity._id}`)
    },
    'toHome': () => {
      return push('/')
    },
    'toParent': (entity) => {
      return push(`/${parentName}/${entity[parentKey]}`)
    }
  }

  /**
   * Do a redirect if needed
   *
   * @param  {Object} cfg Paremters as follows:
   *   - actions: The actions to append to
   *   - endpointCfg: The endpoint config
   *   - defaultStrategy: The default redire strategy to use (default: undefined)
   *   - redirectArgs: Arguments to provide to the redirectionStrategy
   */
  const doRedirectIfNecessary = (cfg) => {
    const redirectionStrategy = (cfg.endpointCfg.redirectionStrategy !== undefined
      ? cfg.endpointCfg.redirectionStrategy
      : cfg.defaultStrategy)

    if (redirectionStrategy) {
      const redirect = redirectionStrategies[redirectionStrategy](cfg.redirectArgs)
      if (redirect) {
        cfg.actions.push(redirect)
      }
    }
  }

  if (commands['Collection GET']) {
    actions.loadAll = (queryParams) => {
      // only include filteredKeys in querystring
      let queryString = ''
      if (queryParams) {
        queryString = qs.stringify(queryParams)
        if (queryString.length !== 0) {
          queryString = '?' + queryString
        }
      }

      // make request
      return fetchAction({
        next: (res, err) => {
          if (!err) {
            return setAllLocally(res.value)
          }
          // TODO handle error
        },
        url: baseEndpoint + queryString
      })
    }
  }

  if (commands['Collection POST']) {
    const endpointCfg = commands['Collection POST']
    actions.create = (newEntity) => {
      return fetchAction({
        next: (res, err) => {
          if (err) {
            // TODO handle error
          } else {
            const createdEntity = res.value
            const actions = [
              addLocally(createdEntity)
            ]
            doRedirectIfNecessary({
              actions,
              endpointCfg,
              defaultStrategy: 'toEntity',
              redirectArgs: createdEntity
            })
            return actions
          }
        },
        options: {
          body: newEntity,
          method: 'POST'
        },
        url: baseEndpoint
      })
    }
  }

  if (commands['DELETE']) {
    const endpointCfg = commands['DELETE']
    actions.delete = (entity) => {
      return fetchAction({
        next: (res, err) => {
          if (err) {
            // TODO handle error
          } else {
            const actions = [
              deleteLocally(entity._id)
            ]
            doRedirectIfNecessary({
              actions,
              endpointCfg,
              defaultStrategy: 'toParent',
              redirectArgs: entity
            })
            return actions
          }
        },
        options: {
          method: 'DELETE'
        },
        url: `${baseEndpoint}/${entity._id}`
      })
    }
  }

  if (commands['GET']) {
    actions.loadOne = (id) => fetchAction({
      next: (res, err) => {
        if (!err) {
          return setLocally(res.value)
        }
        // TODO handle error
      },
      url: `${baseEndpoint}/${id}`
    })
  }

  if (commands['PUT']) {
    const endpointCfg = commands['PUT']
    actions.update = (entity) => fetchAction({
      next: (res, err) => {
        if (err) {
          // TODO handle error
        } else {
          const updatedEntity = res.value
          const actions = [
            setLocally(updatedEntity)
          ]
          doRedirectIfNecessary({
            actions,
            endpointCfg,
            defaultStrategy: 'toEntity',
            redirectArgs: updatedEntity
          })
          return actions
        }
      },
      options: {
        body: entity,
        method: 'PUT'
      },
      url: `${baseEndpoint}/${entity._id}`
    })
  }

  return actions
}

export default makeGenericModelActions
