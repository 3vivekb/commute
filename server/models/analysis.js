const {Schema} = require('mongoose')

const db = require('../db')
const Commuter = db.model('Commuter', require('./commuter'))
const Group = db.model('Group', require('./group'))
const trashPlugin = require('./plugins/trash')
const Site = db.model('Site', require('./site'))
const Trip = db.model('Trip', require('./trip'))
const dbUtils = require('../utils/db')
const later = require('../utils/later')

const schema = new Schema({
  calculationStatus: {
    default: 'calculating',
    required: true,
    type: String
  },
  groupId: {
    ref: 'Group',
    required: true,
    type: Schema.Types.ObjectId
  },
  name: {
    required: true,
    type: String
  },
  numCommuters: {
    default: 0,
    required: true,
    type: Number
  },
  organizationId: {
    ref: 'Organization',
    required: true,
    type: Schema.Types.ObjectId
  },
  siteId: {
    ref: 'Site',
    required: true,
    type: Schema.Types.ObjectId
  },
  summary: {
    // fields
    avgTravelTime: Number,
    avgDistance: Number,
    savingsPerTrip: Number,
    savingsPerTripYear: Number,
    savingsTotalPerDay: Number,
    savingsTotalPerYear: Number,

    // mongoose
    default: {
      avgTravelTime: 0,
      avgDistance: 0,
      savingsPerTrip: 0,
      savingsPerTripYear: 0,
      savingsTotalPerDay: 0,
      savingsTotalPerYear: 0
    },
    required: true,
    type: Schema.Types.Mixed
  }
})

schema.plugin(trashPlugin)

schema.pre('save', dbUtils.makeCascadeDeleteModelsFn('analysisId', [Trip]))

schema.pre('save', true, function (next, done) {
  next()

  const self = this

  if (['skipCalculation', 'calculated'].indexOf(this.calculationStatus) !== -1 ||
    this.trashed) return done()

  const profiler = require('../utils/profiler')

  // initiate profile analysis of all commuters in group
  // find all commuters via groupId
  // make sure group isn't trashed
  const ensureGroupExists = Group.count({ _id: this.groupId, trashed: undefined }).exec()
  const findCommuters = Commuter.find({ groupId: this.groupId, trashed: undefined }).exec()
  const getSite = Site.findOne({ _id: this.siteId, trashed: undefined }).exec()
  Promise.all([findCommuters, getSite, ensureGroupExists])
    .then((results) => {
      // verify that group, commuters and site exist
      const commuters = results[0]
      const site = results[1]
      const groupCount = results[2]

      if (commuters.length === 0) return done('No commuters found!')
      if (!site) return done('Site not found!')
      if (groupCount === 0) return done('Group not found!')

      self.numCommuters = commuters.length

      later(() => {
        profiler({
          analysisId: self._id,
          commuters,
          site
        })
      })
      done()
    })
    .catch((err) => {
      console.error(err)
      self.calculationStatus = 'error'
      done()
    })
})

module.exports = schema
