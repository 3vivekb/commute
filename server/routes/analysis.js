const models = require('../models')
const makeRestEndpoints = require('../utils/restEndpoints')

module.exports = function makeRoutes (app) {
  makeRestEndpoints(app,
    {
      childModels: [{
        foreignKey: 'anlaysisId',
        key: 'trips',
        model: models.Trip
      }],
      commands: {
        'Collection GET': {},
        'Collection POST': {},
        'GET': {},
        'DELETE': {}
      },
      model: models.Analysis,
      name: 'analysis'
    }
  )
}
