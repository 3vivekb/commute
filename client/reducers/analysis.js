import update from 'react-addons-update'

import analysisDefaults from '../utils/analysisDefaults'
import {fixedRound} from '../utils/common'
import {addToEntityMap} from '../utils/entities'
import {makeGenericReducerHandlers} from '../utils/reducers'

const METRICS = Object.keys(analysisDefaults.metrics)

export const reducers = makeGenericReducerHandlers({
  handlers: ['add', 'delete', 'set many'],
  name: {
    singular: 'analysis',
    plural: 'analyses'
  }
})

reducers['receive mock calculated trips'] = function (state, action) {
  const {analysisId, trips} = action.payload
  const affectedAnalysis = state[analysisId]
  let updatedAnalysis = update(affectedAnalysis, {
    trips: { $set: trips }
  })

  // calculate array of values for each metric
  const vals = {}
  const modes = Object.keys(trips[0]).filter((mode) => (['commuterId']).indexOf(mode) === -1)
  modes.forEach((mode) => {
    vals[mode] = {
      cost: [],
      distance: [],
      time: []
    }
  })
  trips.forEach((trip) =>
    modes.forEach((mode) =>
      METRICS.forEach((metric) => {
        vals[mode][metric].push(trip[mode][metric])
      })
    )
  )

  // sort arrays
  modes.forEach((mode) =>
    METRICS.forEach((metric) => {
      vals[mode][metric].sort()
    })
  )

  updatedAnalysis = update(updatedAnalysis, {
    tripVals: { $set: vals }
  })

  // calculate analysis summary metrics
  // Potential Savings is an overall average of the difference in cost between
  // the most likely trip and the driving trip
  const len = trips.length
  let totalCostDifference = 0
  let totalTravelTime = 0
  let totalDistance = 0
  trips.forEach((trip) => {
    totalCostDifference += trip.car.cost - trip.mostLikely.cost
    totalTravelTime += trip.mostLikely.time
    totalDistance += trip.mostLikely.distance
  })

  updatedAnalysis = update(updatedAnalysis, {
    summary: {
      $set: {
        avgTravelTime: fixedRound(totalTravelTime / len),
        avgDistance: fixedRound(totalDistance / len),
        savingsPerTrip: fixedRound(totalCostDifference / len),
        savingsPerTripYear: fixedRound(totalCostDifference / len * 260),
        savingsTotalPerDay: fixedRound(totalCostDifference),
        savingsTotalPerYear: fixedRound(totalCostDifference * 260)
      }
    }
  })

  return addToEntityMap(state, updatedAnalysis)
}

export const initialState = {}
