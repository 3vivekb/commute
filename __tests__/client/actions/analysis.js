/* global describe, it */

import {expectCreateAction, expectDeleteAnalysis} from '../../test-utils/actions'

import * as analysis from '../../../client/actions/analysis'

describe('actions > analysis', () => {
  it('create analysis should work', () => {
    const data = {
      groupId: '1',
      name: 'Mock Analysis',
      siteId: '1'
    }
    const result = analysis.createAnalysis(data, '1')

    expectCreateAction(result)
  })

  it('delete analysis should work', () => {
    const result = analysis.deleteAnalysis('1', '1')

    expectDeleteAnalysis(result)
  })
})
