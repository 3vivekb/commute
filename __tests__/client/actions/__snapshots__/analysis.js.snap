exports[`actions > analysis create analysis should work 1`] = `
Object {
  "payload": Object {
    "groupId": "1",
    "name": "Mock Analysis",
    "siteId": "1",
    "trips": Array [],
  },
  "type": "add analysis",
}
`;

exports[`actions > analysis create analysis should work 2`] = `
Object {
  "payload": Object {
    "args": Array [
      "/organizations/1/analysis/new-entity-id",
    ],
    "method": "push",
  },
  "type": "@@router/CALL_HISTORY_METHOD",
}
`;

exports[`actions > analysis delete analysis should work 1`] = `
Array [
  Object {
    "payload": "1",
    "type": "delete analysis",
  },
  Object {
    "payload": Object {
      "args": Array [
        "/organizations/1",
      ],
      "method": "push",
    },
    "type": "@@router/CALL_HISTORY_METHOD",
  },
]
`;
